File: GoalDrivenPro/GoalDrivenPro/Goal_Think.cpp
#include "Goal_Think.h"
#include "Goal_DevGame.h"
#include "Goal_CodingTask.h"
#include "DevGameGoal_Evaluator.h"



Goal_Think::Goal_Think(Agent* agent) :
	Goal_Composite<Agent>(agent, goal_think)
{
	/*m_Evaluators.push_back(new BuildHealthGoal_Evaluator(1.0));
	m_Evaluators.push_back(new EarnMoneyGoal_Evaluator(1.0));*/

	m_Evaluators.push_back(new DevGameGoal_Evaluator(1.0));
}

Goal_Think::~Goal_Think()
{
	GoalEvaluators::iterator it = m_Evaluators.begin();
	for (; it != m_Evaluators.end(); ++it)
	{
		if (*it)
		{
			delete *it;
		}
	}

	m_Evaluators.clear();
}

void Goal_Think::activate()
{
	if (true)
	{
		arbitrate();
	}

	m_Status = active;
}

int Goal_Think::process()
{
	activateIfInactive();

	int substatus = processSubgoals();

	if (completed == substatus ||
		failed == substatus)
	{
		if (true)
		{
			m_Status = inactive;
		}
	}

	return m_Status;
}

void Goal_Think::terminate()
{

}

void Goal_Think::arbitrate()
{
	double best = 0.0;
	Goal_Evaluator* mostDesirable = nullptr;

	GoalEvaluators::iterator curDes = m_Evaluators.begin();
	for (curDes; curDes != m_Evaluators.end(); ++curDes)
	{
		double desirability = (*curDes)->calculateDesirability(m_Owner);

		if (desirability >= best)
		{
			best = desirability;
			mostDesirable = *curDes;
		}
	}

	if (mostDesirable) mostDesirable->setGoal(m_Owner);
}

bool Goal_Think::handleMessage()
{
	return true;
}

void Goal_Think::addGameDevGoal()
{
	addSubgoal(new Goal_DevGame(m_Owner));
}

void Goal_Think::addCodingTask()
{
	addSubgoal(new Goal_CodingTask(m_Owner));
}

File: Goal_driven/Goal_Think.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Chuizi.Goal_driven
{
    class Goal_Think : Goal_Composite<Agent>
    {
        private List<Goal_Evaluator> GoalEvaluators =
            new List<Goal_Evaluator>();

        public Goal_Think(Agent ag) : base(ag, Agent_Goal_Types.goal_think)
        {
            const double LowRangeOfBias = 0.5;
            const double HightRangeOfBias = 1.5;

            double moneyBias = Math.Round(HightRangeOfBias - LowRangeOfBias) + LowRangeOfBias;
            double willingGirlsBias = Math.Round(HightRangeOfBias - LowRangeOfBias) + LowRangeOfBias;
            double studyBias = Math.Round(HightRangeOfBias - LowRangeOfBias) + LowRangeOfBias;

            //GoalEvaluators.Add(new GetMoney_Evaluator(moneyBias));
            GoalEvaluators.Add(new GetGirls_Evaluator(willingGirlsBias));
            GoalEvaluators.Add(new Study_Evaluator(studyBias));
        }

        ~Goal_Think()
        {
            GoalEvaluators.Clear();
        }

        public void Arbitrate()
        {
            double best = 0;
            Goal_Evaluator mostDesirable = null;

            foreach (var item in GoalEvaluators)
            {
                double desirability = item.CalculateDesirability(Owner);
                if (desirability >= best)
                {
                    best = desirability;
                    mostDesirable = item;
                }
            }

            if (mostDesirable != null) {
                mostDesirable.SetGoal(Owner);
            }
        }

        public bool notPresent(Agent_Goal_Types goaltype)
        {
            if (SubGoals.Count > 0)
            {
                return SubGoals.Peek().GetGoalType() != goaltype;
            }

            return true;
        }

        public override Goal_Status Process()
        {
            ActiveIfInactive();

            Goal_Status sgs = ProcessSubgoals();

            if (sgs == Goal_Status.completed || sgs == Goal_Status.failed)
            {
                //if (!Owner.isProcessed())
                {
                    Status = Goal_Status.inactive;
                }
            }

            return Status;
        }

        public override void Activate()
        {
            //if (!Owner.isProcessed())
            {
                Arbitrate();
            }

            Status = Goal_Status.active;
        }

        public override void Terminate()
        {
            base.Terminate();
        }

        public void AddGoal_BuildProducts()
        {
            if (notPresent(Agent_Goal_Types.goal_buildProduct))
            {
                RemoveAllSubgoals();
            }
        }

        public void AddGoal_EarnMoney()
        {
            AddSubgoal(new Goal_GetMoney(Owner));
        }

        public void AddGoal_PlayWithGirls()
        {
            AddSubgoal(new Goal_GetGirls(Owner));
        }

        public void AddGoal_Study()
        {
            AddSubgoal(new Goal_Study(Owner));
        }
    }
}
